### PR TITLE
Add GpuEqualToNoNans and update GpuPivotFirst to use to handle PivotFirst with NaN support enabled on GPU

### LINKED ISSUE #5322: [FEA] remove hasNans for Pivot
**Is your feature request related to a problem? Please describe.**

Spark is inconsistent in that for pivot NaN != NaN, but for our code it is. This is because we are using a GpuEqualTo internally to check if the value matches the pivot column, and GpuEqualTo has fixes in it to make NaN == NaN. We can fix this with an alternative GpuEqualTo that removes the NaN == NaN fixes.

### PR DESCRIPTION
**Pull request by NVnavkumar**
Fixes #5322.

This adds the GpuEqualToNoNans operation to support EqualTo where NaN != NaN. This is then used in GpuPivotFirst to allow GPU and CPU behavior to be consistent when the plugin knows that floats can be NaN. The check for the hasNans configuration is now removed so the operation will continue to run on the GPU even if the input data has NaN values.  The integration tests have also been updated to account for `hasNans` and support NaN float inputs in `pivot(...)` on the GPU.

Signed-off-by: Navin Kumar <navink@nvidia.com>

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala [hunk 1]
```diff
@@ -316,6 +316,79 @@ case class GpuEqualNullSafe(left: Expression, right: Expression) extends CudfBin
...
 }
 
+case class GpuEqualToNoNans(left: Expression, right: Expression) extends CudfBinaryComparison
+    with NullIntolerant {
+  override def symbol: String = "="
+  override def outputTypeOverride: DType = DType.BOOL8
+  override def binaryOp: BinaryOp = BinaryOp.EQUAL
+
+  override def doColumnar(lhs: GpuColumnVector, rhs: GpuColumnVector): ColumnVector = {
+    val result = super.doColumnar(lhs, rhs)
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala:326]: Does CUDF compare NaNs as equal? I thought it didn't because C++ does not, because it follow the ieee standard.

>    The comparisons EQ, GT, GE, LT, and LE, when either or both operands is NaN returns FALSE.
>
>    The comparison NE, when either or both operands is NaN returns TRUE.

I htink we can deleted all of the overrides for doColumnar and just rely on the default implementations.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala [hunk 2]
```diff
@@ -316,6 +316,79 @@ case class GpuEqualNullSafe(left: Expression, right: Expression) extends CudfBin
...
+            }
+          }
+        }
+      }
+    } else {
+      result
+    }
+  }
+
+  override def convertToAst(numFirstTableColumns: Int): ast.AstExpression = {
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala:382]: I don't think we have a way to reach this code currently. We don't support AST in aggregation projections like that, so I think it would be cleaner to just delete this method and let the default implementation throw an exception if it is called.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala [hunk 3]
```diff
@@ -316,6 +316,79 @@ case class GpuEqualNullSafe(left: Expression, right: Expression) extends CudfBin
   }
 }
 
+case class GpuEqualToNoNans(left: Expression, right: Expression) extends CudfBinaryComparison
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala:319]: nit: Can we add a few scaladocs to this to explain why we need this?

**Review by revans2 — COMMENTED** [review]
Looks good, I just would like to see if we can make it smaller.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala (hunk 1 above)
Comment (NVnavkumar) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala]: removed doColumnar overrides in last commit

**Comment by NVnavkumar** [comment]
build

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 4]
```diff
@@ -510,41 +488,47 @@ def test_hash_multiple_grpby_pivot(data_gen, conf):
...
         conf=conf)
 
+@approximate_float
+@ignore_order(local=True)
+@allow_non_gpu('HashAggregateExec', 'PivotFirst', 'AggregateExpression', 'Alias', 'GetArrayItem',
+        'Literal', 'ShuffleExchangeExec', 'HashPartitioning', 'KnownFloatingPointNormalized',
+        'NormalizeNaNAndZero')
+@incompat
+@pytest.mark.parametrize('data_gen', [_grpkey_floats_with_nulls_and_nans], ids=idfn)
+def test_hash_pivot_groupby_duplicates_fallback(data_gen):
```
Comment (nartal1) [inline integration_tests/src/main/python/hash_aggregate_test.py:522]: Thanks for adding this test. I had documented in GpuOverrides but forgot to add the test when pivot values has duplicates.

**Review by revans2 — APPROVED** [review]
